pandas/core/resample.py

```diff
@@ -1357,20 +1365,51 @@ def __init__(
         self.convention = convention or "E"
         self.convention = self.convention.lower()
 
-        if isinstance(loffset, str):
-            loffset = to_offset(loffset)
-        self.loffset = loffset
-
         self.how = how
         self.fill_method = fill_method
         self.limit = limit
-        self.base = base
+
+        self.origin = Timestamp(origin) if origin is not None else None
+        self.offset = Timedelta(offset) if offset is not None else None
 
         # always sort time groupers
         kwargs["sort"] = True
 
+        self._warn_if_loffset_or_base_is_used(base, loffset, _warning_stack_level)
+        if isinstance(loffset, str):
+            loffset = to_offset(loffset)
+        self.loffset = loffset
+        self.base = base if base else 0
+
         super().__init__(freq=freq, axis=axis, **kwargs)
 
+    @staticmethod
```

Comment:
don't use static methods, you can create a module level function if you really need to, but I am not sure why you don't just in-line the warning (is it emitted in more than 1 place?)

Reply:
This should be resolved in the last commit.
> I am not sure why you don't just in-line the warning (is it emitted in more than 1 place?)

Well, yes and no... technically the TimeGrouper can raise alone the warnings. But the stack_level will be wrong in some cases depending if TimeGrouper is called from `.resample` (in `groupby.py` and `generic.py`) or from `pd.Grouper`... Here why I have added the nasty `_warning_stack_level` (a solution that I did not really liked to be honest)

Now I'm calling the function `_validate_resample_deprecated_args` from `__new__` of pd.Grouper or from the two `.resample` functions in `generic.py` or in `groupby.py`.

Reply:
Actually I found a way (but I don't know if I should do it or not) to `in-line` the warning. It can only be done in the `__new__` method of `pd.Grouper`. :man_shrugging: 

Depending if the `cls` is `TimeGrouper` (internal call of pandas) or `Grouper` I can infer what to do. Here a snippet of code that I just tested. Let me know if you prefer the manual call of `_validate_resample_deprecated_args` from `resample` or from `pd.Grouper` or this solution.

```python
class Grouper:

    def __new__(cls, *args, **kwargs):
        if kwargs.get("freq") is not None:
            from pandas.core.resample import TimeGrouper

            # if cls is TimeGrouper that means that the call comes from the
            # pandas internal of resample, otherwise it comes from pd.Grouper,
            # hence this test.
            warn_level = 4 if cls is TimeGrouper else 2
            if kwargs.get("base", None) is not None:
                warnings.warn(
                    "'base' in .resample() and in Grouper() is deprecated.\n"
                    "The new arguments that you should use are 'offset' or 'origin'.\n\n"
                    '>>> df.resample(freq="3s", base=2)\n'
                    "\nbecomes:\n\n"
                    '>>> df.resample(freq="3s", offset="2s")\n',
                    FutureWarning,
                    stacklevel=warn_level,
                )
                if kwargs.get("offset", None) is not None:
                    raise ValueError("offset and base cannot be present at the same time")

            if kwargs.get("loffset", None) is not None:
                warnings.warn(
                    "'loffset' in .resample() and in Grouper() is deprecated.\n"
                    "Here an example to have the same behavior than loffset:\n\n"
                    '>>> df.resample(freq="3s", loffset="8H")\n'
                    "\nbecomes:\n\n"
                    ">>> from pandas.tseries.frequencies import to_offset\n"
                    '>>> df = df.resample(freq="3s").mean()\n'
                    '>>> df.index = df.index.to_timestamp() + to_offset("8H")\n',
                    FutureWarning,
                    stacklevel=warn_level,
                )

            cls = TimeGrouper
        return super().__new__(cls)
```